src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -170,7 +171,13 @@ String getClassNameInternal() {
         cacheKey = new CityHash128Key(sql);
 
         // Parse or fetch SQL metadata from cache.
-        ParsedSQLCacheItem parsedSql = getOrCreateCachedParsedSQL(cacheKey, sql);
+        ParsedSQLCacheItem parsedSql = getCachedParsedSQL(cacheKey, sql);
+        if (null != parsedSql) {
+            isExecutedAtLeastOnce = true;
+        }
+        else {
+            parsedSql = parseAndCacheSQL(cacheKey, sql);
+        }
```

Comment:
@TobiasSQL Hang on a sec. 😁   This is a JDBC driver, right?  I think if you survey the landscape you'll find that ``PreparedStatements`` with identical SQL are *overwhelmingly* reused across connections, by a large margin.

JEE containers, Spring containers, web containers (Tomcat/Jetty), JPA persistence, almost without fail acquire a connection from a pool, execute some business logic involving static (but parameterized) SQL, via ``PreparedStatement``, and return the connection to the pool.

Skipping *sp_executesql* after observing multiple executions of identical SQL is not the same as skipping *sp_prepsql* and going directly to *sp_execute*.  Which by the way is what my ``speedify`` branch does, because it caches *input* parameter metadata and goes directly to *sp_execute* for identical SQL with identical parameter types, additionally using the *fReuseMetaData* flag (and *fNoMetaData*, because the *column info* metadata coming out is cached as well).<sup>[reference](https://msdn.microsoft.com/en-us/library/dd357576.aspx)</sup>

Functionally, using *sp_executesql* within the ``PreparedStatement`` is purely optional, but *I do* recognize the case for *dynamic* SQL executed via ``PreparedStatement``.  If it is truly dynamic, the behavior above is exactly as you would expect/want.  If it is *quasi-dynamic* wherein identical SQL *is* being executed with some frequency, I would argue that it would benefit from the optimizations that (will) come with *fReuseMetaData* and *fNoMetaData* by skipping the *sp_executesql* (and even the *sp_prepsql* when those changes land).

Keep in mind that "*is being executed with some frequency*" means multiple executions of identical SQL within the timeframe in which parsed SQL has not rolled out of the cache.  With a *lot* of (truly) dynamic SQL the cache is going to roll very often.  We talking about *identical* SQL executed *within a narrow time horizon* being "promoted" to *sp_prepsql* (regardless of connection affiliation).